src/posts/posts.service.ts

```diff
@@ -0,0 +1,132 @@
+import {
+    BadRequestException,
+    ForbiddenException,
+    Injectable,
+    NotFoundException,
+} from '@nestjs/common';
+import { CreatePostDto } from './dto/create-post.dto';
+import { InjectRepository } from '@nestjs/typeorm';
+import { PostEntity } from './post.entity';
+import { Repository } from 'typeorm';
+import { UserEntity } from '../users/user.entity';
+import { FilesService } from '../files/files.service';
+import { FileDto } from '../files/dto/file.dto';
+import { IPaginationOptions, Pagination, paginate } from 'nestjs-typeorm-paginate';
+import { Order } from '../../common/enums/pagination.order';
+import { OrderCondition } from '../../common/enums/order.condition';
+import { UpdatePostDto } from './dto/update-post.dto';
+
+@Injectable()
+export class PostsService {
+    constructor(
+        @InjectRepository(PostEntity) private postsRepo: Repository<PostEntity>,
+        private filesService: FilesService,
+    ) {}
+
+    async createPost(
+        createPostDto: CreatePostDto,
+        user: UserEntity,
+        file?: Express.Multer.File,
+    ): Promise<PostEntity> {
+        let createFileDto: FileDto;
+        if (file) {
+            createFileDto = await this.filesService.uploadFile(file.buffer, file.originalname);
+            createPostDto.imageKey = createFileDto.key;
+        }
+        const post = await this.postsRepo.create({
+            ...createPostDto,
+            user: { id: user.id },
+        });
+        const createdPost = await this.postsRepo.save(post);
+        if (createFileDto) {
+            await this.filesService.createFile(createFileDto, createdPost.id);
+        }
+        return createdPost;
+    }
+
+    async findAll(
+        options: IPaginationOptions,
+        order: Order,
+        condition: OrderCondition,
+    ): Promise<Pagination<PostEntity>> {
+        const queryBuilder = await this.postsRepo.createQueryBuilder('post');
+        queryBuilder.orderBy(`post.${condition}`, order);
+        return paginate<PostEntity>(queryBuilder, options);
+    }
+
+    async findOne(id: number) {
+        return await this.postsRepo.findOne({ where: { id } });
+    }
+
+    async getImage(id: number): Promise<any> {
+        const post = await this.postsRepo.findOne({ where: { id }, relations: ['file'] });
+        if (!post) {
+            throw new NotFoundException('Post not found');
+        }
+        if (!post.file) {
+            throw new NotFoundException('There is no image in this post');
+        }
+        return this.filesService.downloadFile(post.file.key);
+    }
+
+    async deleteImage(id: number, currentUser: UserEntity) {
+        const post = await this.postsRepo.findOne({ where: { id }, relations: ['file', 'user'] });
+        if (!post) {
+            throw new NotFoundException('Post not found');
+        }
+        if (currentUser.role !== 'admin' && currentUser.id !== post.user.id) {
```

Comment:
Please don't use hardcoded value, use `Role.Admin` instead of `'admin'`.

Reply:
You can move this validation to the line: 73. It is more optimal to check the permission and only then make a DB request to get the post.